### PR TITLE
Different data for unary calls

### PR DESCRIPTION
**Pull request by ezsilmar**
This PR addresses issue https://github.com/bojand/ghz/issues/82

Changes sum-up:
1. If multiple messages are specified for an unary call they are sent in a round-robin fashion.
2. Binary format now supports multiple messages with count-delimited encoding as per [this doc](https://developers.google.com/protocol-buffers/docs/techniques#streaming).
3. Binary messages are cached for the worker and not parsed on every request. 
4. Docs are updated to reflect the changes.
5. Functional tests now have a random delay in every handler to make time-checks more stable.

### CODE REVIEW

File: runner/worker.go [hunk 1]
```diff
@@ -106,20 +93,57 @@ func (w *Worker) makeRequest() error {
...
+	if !w.config.binary {
+		data, err := ctd.executeData(string(w.config.data))
+		if err != nil {
+			return nil, err
+		}
+		inputs, err = createPayloadsFromJSON(string(data), w.mtd)
+		if err != nil {
+			return nil, err
+		}
+		// Json messages are not cached due to templating
```
Comment (bojand) [inline runner/worker.go]: 👍

**Comment by ezsilmar** [comment]
Hello, wow, didn't know that `usage.md` is auto-generated.

Thank you for the tool :)

**Comment by egick666** [comment]
That's a really useful feature! Thank you guys for that! It's really just in time for me.

I'm wondering on resource requirements for machine running ghz tool. 
I'm asking because I've tried to run some traffic using unary calls with  request data array consists of 1000 messages using several clients with quite small qps and reached the top of CPU usage immediately.
I'm using VM with 4vCPU and Xeon(R) CPU X5670 @ 2.93GHz installed on my machine.
Have you tried to run more load in round-robin unary call mode?

**Comment by ezsilmar** [comment]
Do you have your data in JSON format or binary? 

Binary format is much more performant in multiple messages scenario because messages are parsed once and then cached. On Xeon E3-1271 with 50 binary messages (data file is 120 KB) I hit ~6000 qps (server-side bottleneck) and client cpu usage doesn't go above 17%.

**Comment by egick666** [comment]
I'm using JSON format.

If you could suggest the way how can I convert JSON data file into binary one and use with ghz I would be happy ;)

**Comment by ezsilmar** [comment]
You just need to serialize your grpc call parameter directly to a file using any protobuf library (language doesn't matter). To serialize several messages prefix every message with a number of bytes it occupies. Most libraries have a special method for this.

Then pass the file to ghz with `-B` option. There's an example available in the documentation: https://ghz.sh/docs/options#b-binary

Hope this will help!

**Comment by bojand** [comment]
Hello, thanks for the discussion and for the useful tips @ezsilmar ! 

@egick666 With regards to the CPU usage, there are many factors that could be at play. Is the server also running locally? What are the test parameters, more specifically the total number of requests and the concurrency used? Are you limiting the # of cores using [`--cpus`](https://ghz.sh/docs/options#cpus) or some other means? Also if you're running other applications (ie. Slack and other Electron apps that are resource-hungry) that may contribute to cpu usage. Also `1000` messages is a lot of data but I would expect that more to contribute to memory usage rather than CPU. Just some things to look into and play around I guess. Finally CPU @ 100% for duration test is not necessarily that bad in some ways. We want the computer to be used to its full efficiency potential heh. But I understand the concern.

**Comment by egick666** [comment]
Hello @bojand.
I'm running only ghz on this VM without any other software. Traffic is running for a particular period of time using -z parameter specified instead of -n. Concurrency is -c 8 with 8 connections, no limit for CPUs.

Having CPU @ about 100% make client as a bottleneck and doesn't allow me to gather clear performance numbers of my gRPC server.

@ezsilmar 
Do you have any examples of request serialization based on proto file with prepending it with the message size? Would be great to take a look, just to make sure that I understand what you are talking about.

Thanks!

**Comment by bojand** [comment]
Hello, thanks for the details. Definitely seems related to the # of items in data. From a quick test I am able to reproduce locally with 1000 items in an array of data for a unary call. Seems to work fine with a few hundred items but not with 1000. Although in my testing it did not seem to have any impact on actual results from the server, but this was a pretty simple quick test using the greeter gRPC example. I see some room for optimization when we are dealing with larger data sets. Can you please create a new issue. Make sure to include as much details as possible, specifically: the type of call, the VM aspect of test and how are you measuring the CPU, and all the test parameters.

Thanks.

**Comment by egick666** [comment]
Thanks @bojand for explanation and example! I'm not really familiar with Golang, but your comments helped. I'll try to run the same traffic in binary format and get back to you with results.

I'll create new issue for JSON and pass details there. Hope it'll help to make this feature usefull for JSON as well.

**Pull request merged**